### PR TITLE
N-Distance nodes & edges

### LINKED ISSUE #3248: Active Node-Edge N-Distance GraphQL Query
This is in reference to the desired HYBRID mode noted in the Exeter-June-2019 meeting, and as a solution could present a way of masking the Cylc 8 to 9 backend transition.

To create a query to return that active task/job-edge view:
- Define a type (or query with fields) whose fields are a node/task-proxy list and list of edges
- Define query arguments with one filtering on node atts (i.e. active states for current data provision), and another `level` or `distance` (i.e. geodesic distance, because level implies a tree?).
- Create a resolver to collect the selected  `D0` (`distance`=0) nodes, then loop on all edges to collect edges and nodes (`D1`) who are the source or target of `D0`... Repeat for `D1..N`

If we think that resolver would be too computationally expensive, then for every node we could store edges that connect to it, as either source or target, in a list field of that node.
(by store I mean the edge IDs of course)

### PR DESCRIPTION
**Pull request by dwsutherland**
These changes close #3248 
Accompanying UI Server PR: https://github.com/cylc/cylc-uiserver/pull/70

**Changes:**
- **`NodesEdges`:** type and query added to the GraphQL schema.
- **`edges` field added to the `PbTaskProxy` nodes:** for efficient resolution of the `nodesEdges` query.
- **Reformed data-store:** This was done to have a uniform access for the resolvers at both WS an UIS.
- **Faster resolvers!** All `resolvers` now use the native data-store map structure, instead of the `entire_workflow` protobuf message (attributes and lists), so all resolvers that use `id` to directly retrieve data (not root queries) should be faster (in theory).
- **`BaseResolvers` class created:** Contains all resolvers for import and/or use by both WS an UIS, with local `Resolvers` class inheriting this.
- **`get_nodes_edges` resolver created:** (in `BaseResolvers`) to gather the data elements for `nodesEdges` query resolution.

**Resolver & Query**
The resolver works by filtering for root or starting nodes with similar filter args as `taskProxies`:
image

the one new argument is `distance` which, inline with [graph theory](https://en.wikipedia.org/wiki/Distance_(graph_theory)), is the number of edges (in shortest path) between two nodes... To the resolver, `distance` is how many iterations of the following:
- Start with `root/new nodes` (distance `n-1`).
- Use the edges field of the nodes to acquire `new edges`.
- Add `new edges` to `edges` (edges should be unique).
- Gather the source & target nodes from these edges, the difference of these and current set of nodes give `new nodes` (nodes should be unique after use of difference).
- Add `new nodes` to `nodes`.
- Repeat to `n`

**Notes:** `distance` defaults to 1, a value of zero (or negative values) return the root nodes only, the algorithm will keep going until either:
- there are no more new nodes
- `distance` is reached

Because sets and set operations are used, then when new edges meet, the search won't overlap.. Only new node edges are used...

The GraphQL query looks like:
```
query {
  nodesEdges(
      workflows: ["*|*"],
  	  ids: ["*|*"],
      states: ["running", "succeeded"],
      distance: 0,
      sort: {keys: ["cycle_point"], reverse: false}) {
    nodes {
      name
      cyclePoint
      state
      isHeld
    }
    edges {
      id
    }
  }
}
```
(nodes having `taskProxy` fields)

**Example:**
Take the following suite:
```
[cylc]
    cycle point format = CCYYMMDDThh
    UTC mode = True
[scheduling]
    initial cycle point = 20190101T00
    max active cycle points = 2
    [[graph]]
        P1D = """
foo[-P1D] => foo
foo => baa => bar => baz  # long chain
qux  # isolate
"""
[runtime]
    [[root]]
        script = sleep 15; echo "$CYLC_TASK_NAME"
    [[qux]]
        script = sleep 60; echo "$CYLC_TASK_NAME"
```
The above query yields
```
{
  "data": {
    "nodesEdges": {
      "nodes": [
        {
          "name": "foo",
          "cyclePoint": "20190101T00",
          "state": "running"
        },
        {
          "name": "qux",
          "cyclePoint": "20190101T00",
          "state": "running"
        },
        {
          "name": "baa",
          "cyclePoint": "20190101T00",
          "state": "waiting"
        },
        {
          "name": "qux",
          "cyclePoint": "20190102T00",
          "state": "running"
        },
        {
          "name": "foo",
          "cyclePoint": "20190102T00",
          "state": "waiting"
        }
      ],
      "edges": [
        {
          "id": "sutherlander|noz|foo.20190101T00|foo.20190102T00"
        },
        {
          "id": "sutherlander|noz|foo.20190101T00|baa.20190101T00"
        }
      ]
    }
  }
}
```

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Already covered with extension of existing tests.
- [x] No change log entry required (invisible to users).
- [x] No documentation update required.

### CODE REVIEW

**Comment by kinow** [comment]
Should we care about negative or zero distance? I tested to make sure the UI wouldn't crash, but it seems like it simply ignores the distance when `0` or `-1` are given as argument to `nodeEdges(distance: -1){}`.

**Comment by kinow** [comment]
Oh, would be nice to have some docs in GraphiQL too. Here's the `maxDepth documentation for `Tasks`:

image

But some fields in `Edge` - for instance - are returning "No Description", e.g. `Edge.targetNode`:

image

And `Edge.cond` and `Edge.suicide` have both the same documentation "The Boolean scalar type represents true or false".

image

Would be nice to have some text describing what the flag does/enables/disables/etc.

Thanks!

File: cylc/flow/network/resolvers.py [hunk 1]
```diff
@@ -204,20 +217,85 @@ def __init__(self, schd):
...
             args)
 
+    async def get_nodes_edges(self, root_nodes, args):
+        """Return nodes and edges within a specified distance of root nodes."""
+        # Initial root node selection.
+        node_ids = set(n.id for n in root_nodes)
+        edge_ids = set()
+        # Setup for edgewise search.
+        new_nodes = root_nodes
+        new_node_ids = node_ids
```
[the remark this answers is not in the document]
Comment (dwsutherland) [reply cylc/flow/network/resolvers.py]: Thing is, we want root nodes and no edges for `distance=0` and a set of edges and corresponding edges for every distance thereafter... I'll welcome suggestions though.

**Comment by kinow** [comment]
@dwsutherland, moved the part about documenting the returned data to: https://github.com/cylc/cylc-flow/issues/3340

**Comment by dwsutherland** [comment]
@kinow 

> 1. `suicide` I believe is a boolean flag to say whether the node is representing a suicide task or not. What is the `cond` boolean for? The GraphiQL interface gives no description for that field.

Those are edge attributes (not node) .. So they describe something about the nature of the dependency between source and target nodes. (conditional relationship `foo | bar => qux` , and suicide `foo => !bar`).. They were in the `get_graph_raw` edges so didn't take them out.

> 2. What about `targetNode` and `sourceNode`? They appear to only contain `__typename` attribute, which will always be `TaskProxy` I guess?

They are `fragment` fields, because originally the nodes could be family or task proxies... So you can do this:

```
read -r -d '' nodesEdges <<_args_
{
  "request_string": "
fragment tProxy on TaskProxy {
  state
  cyclePoint
  latestMessage
  jobs {
    id
    state
    host
    batchSysName
    batchSysJobId
    submittedTime
    startedTime
    finishedTime
    submitNum
  }
}

fragment fProxy on FamilyProxy {
  state
  cyclePoint
}

query {
  nodesEdges(workflows: [\"${1}\"], ids: [\"foo\"], distance: ${DISTANCE}, sort: {keys: [\"state\", \"id\"], reverse: false}) {
    nodes {
      id
      state
    }
    edges {
      id
      source
      sourceNode {
        ...tProxy
        ...fProxy
      }
      target
      targetNode {
        ...tProxy
        ...fProxy
      }
    }
  }
}",
"variables": null
}
_args_
```



> 
> I had a few issues with the Graph.vue, and Cytoscape. But the only issue I had with the data returned, was that the edges returned contained nodes that were not in the list of nodes.
> 
> I hacked - some more - code to filter edges with nodes missing `source` or `target` from the list of `nodesEdges.nodes`.

The only "nodes" that should be missing from the `nodesEdges` query output are the xtriggers....
Otherwise there's a bug..

**Comment by kinow** [comment]
@dwsutherland brilliant! Thanks for the answers! I've still got some more homework to learn more about GraphQL 😞 but getting there!

**Comment by hjoliver** [comment]
The boolean for "conditional dependency" is used to style edges differently (hollow arrow-head) in Cylc 7.

File: cylc/flow/network/resolvers.py [hunk 2]
```diff
@@ -153,47 +167,52 @@ def __init__(self, schd):
...
+                args)
         return sort_elements(
             [node
-             for node in nodes
-             if node.id in nat_ids and node_filter(node, args)],
+             for flow in [self.data[w_id] for w_id in w_ids]
+             for node in [flow[node_type][n_id]
+                          for n_id in nat_ids
+                          if n_id in flow[node_type]]
+             if node_filter(node, args)],
```
Comment (matthewrmshin) [inline cylc/flow/network/resolvers.py]: Any chance of refactoring the block between L174 and L192 so it can be easier to understand?

File: cylc/flow/network/resolvers.py [hunk 3]
```diff
@@ -153,47 +167,52 @@ def __init__(self, schd):
...
         """Return protobuf node object for given id."""
         n_id = args.get('id')
         o_name, w_name, _ = n_id.split(ID_DELIM, 2)
         w_id = f'{o_name}{ID_DELIM}{w_name}'
-        if self.schd.ws_data_mgr.workflow.id != w_id:
+        flow = self.data.get(w_id)
+        if not flow:
             return None
-        flow_msg = self.schd.ws_data_mgr.get_entire_workflow()
         if node_type == 'proxy_nodes':
```
Comment (matthewrmshin) [inline cylc/flow/network/resolvers.py]: Not part of this change, but is it possible to make `'proxy_nodes'` a constant, as it is used in various places.

File: cylc/flow/network/resolvers.py [hunk 4]
```diff
@@ -204,20 +223,91 @@ def __init__(self, schd):
...
+                o_name, w_name, _ = node_id.split(ID_DELIM, 2)
+                w_ids.add(f'{o_name}{ID_DELIM}{w_name}')
+            new_nodes = [
+                node
+                for flow in [self.data[w_id] for w_id in w_ids]
+                for node in [
+                    flow['task_proxies'][n_id]
+                    for n_id in new_node_ids
+                    if n_id in flow['task_proxies']]
+            ]
```
Comment (matthewrmshin) [inline cylc/flow/network/resolvers.py]: This pattern appears in various places?

File: cylc/flow/ws_data_mgr.py [hunk 5]
```diff
@@ -478,30 +488,34 @@ def generate_graph_elements(self, edges=None,
...
+            self.generate_ghost_families(family_proxies, new_points)
 
-        self.generate_ghost_families(family_proxies, new_points)
         # Replace the originals (atomic update, for access from other threads).
-        self.task_proxies = task_proxies
-        self.edges = edges
-        self.graph = graph
+        self.data[self.workflow_id]['task_proxies'] = task_proxies
+        self.data[self.workflow_id]['edges'] = edges
+        self.data[self.workflow_id]['graph'] = graph
```
Comment (matthewrmshin) [inline cylc/flow/ws_data_mgr.py]: Worth using constants for some of these keys?

File: cylc/flow/network/resolvers.py (hunk 4 above)
Comment (dwsutherland) [reply cylc/flow/network/resolvers.py]: Yes it replaced iterating over a list with a hash table lookup.

File: cylc/flow/network/resolvers.py (hunk 2 above)
Comment (dwsutherland) [reply cylc/flow/network/resolvers.py]: List comprehensions aren't the easiest to follow.. Would you like some comments instead? or the
```
 [flow['task_proxies'][n_id]
  for n_id in nat_ids
  if n_id in flow['task_proxies']]
```
block made a function?

File: cylc/flow/ws_data_mgr.py (hunk 5 above)
Comment (dwsutherland) [reply cylc/flow/ws_data_mgr.py]: You mean:
```
TASK_PROXIES='task_proxies'
.
.
.
        self.data[self.workflow_id][TASK_PROXIES] = task_proxies
```
at the class or module level?

File: cylc/flow/network/resolvers.py (hunk 2 above)
Comment (matthewrmshin) [reply cylc/flow/network/resolvers.py]: Fantastic if you can make the block a function - so we don't end up with so many of these list comprehensions (that are difficult to read at least for me).

Comment (oliver-sanders) [reply cylc/flow/network/resolvers.py]: Moving the `node_type` logic out of the list comprehension it becomes pretty clear:

```python
if node_type == PROXY_NODES:
    node_types = [TASK_PROXIES, FAMILY_PROXIES]
else:
    node_types = [node_type]

return sort_elements(
    [
        node
        for flow in get_flow_data(self.data, w_ids)
        for node_type in node_types
        for node in get_data_elements(flow, nat_ids, node_type)
        if node_filter(node, args)
    ],
    args
)
```

Comment (matthewrmshin) [reply cylc/flow/network/resolvers.py]: I guess I'm just not blessed with the talent of reading a complex list comprehension like this one with 3 for-loops and an if-block...

Comment (dwsutherland) [reply cylc/flow/network/resolvers.py]: Done.

**Comment by hjoliver** [comment]
That's three approvals.
